src/battle/overlay_12_0224E4FC.c

```diff
@@ -1962,3 +1965,920 @@ void ov12_02251710(BattleSystem *bsys, BATTLECONTEXT *ctx) {
+        msg->param[1] = ctx->battleMons[battlerId].moves[movePos];
+        ret = FALSE;
+    } else if (StruggleCheck(bsys, ctx, battlerId, 0, STRUGGLE_CHECK_GRAVITY) & MaskOfFlagNo(movePos)) {
+        msg->tag = 10;
+        msg->id = 1001;
+        msg->param[0] = CreateNicknameTag(ctx, battlerId);
+        msg->param[1] = ctx->battleMons[battlerId].moves[movePos];
+        ret = FALSE;
+    } else if (StruggleCheck(bsys, ctx, battlerId, 0, STRUGGLE_CHECK_HEAL_BLOCK) & MaskOfFlagNo(movePos)) {
+        msg->tag = 34;
+        msg->id = 1057;
+        msg->param[0] = CreateNicknameTag(ctx, battlerId);
+        msg->param[1] = MOVE_HEAL_BLOCK;
+        msg->param[2] = ctx->battleMons[battlerId].moves[movePos];
+        ret = FALSE;
+    } else if (StruggleCheck(bsys, ctx, battlerId, 0, STRUGGLE_CHECK_CHOICED) & MaskOfFlagNo(movePos)) {
+        msg->tag = 24;
+        msg->id = 911;
+        msg->param[0] = ctx->battleMons[battlerId].item;
+        msg->param[1] = ctx->battleMons[battlerId].unk88.moveNoChoice;
+        ret = FALSE;
+    } else if (StruggleCheck(bsys, ctx, battlerId, 0, STRUGGLE_CHECK_NO_PP) & MaskOfFlagNo(movePos)) {
+        msg->tag = 0;
+        msg->id = 823;
+        ret = FALSE;
+    } 
+    
+    return ret;
+}
+
+int BattleMon_GetMoveIndex(BATTLEMON *mon, u16 moveNo) {
+    int movePos;
+    
+    for (movePos = 0; movePos < 4; movePos++) {
+        if (mon->moves[movePos] == moveNo) {
+            break;
+        }
+    }
+    
+    return movePos;
+}
+
+extern u8 sTypeEffectiveness[112][3];
+
+//static
+BOOL ov12_02251C74(BATTLECONTEXT *ctx, int battlerIdAttacker, int battlerIdTarget, int index) {
+    int item = GetBattlerHeldItemEffect(ctx, battlerIdTarget);
+    BOOL ret = TRUE;
+    
+    if (item == HOLD_EFFECT_SPEED_DOWN_GROUNDED || (ctx->battleMons[battlerIdTarget].moveEffectFlags & MOVE_EFFECT_INGRAIN)) {
+        if (sTypeEffectiveness[index][1] == TYPE_FLYING && sTypeEffectiveness[index][2] == TYPE_MUL_NO_EFFECT) {
+            ret = FALSE;
+        }
+    }
+    
+    if (ctx->turnData[battlerIdTarget].roostFlag && sTypeEffectiveness[index][1] == TYPE_FLYING) {
+        ret = FALSE;
+    }
+    
+    if (ctx->fieldCondition & FIELD_CONDITION_GRAVITY) {
+        if (sTypeEffectiveness[index][1] == TYPE_FLYING && sTypeEffectiveness[index][2] == TYPE_MUL_NO_EFFECT) {
+            ret = FALSE;
+        }
+    }
+    
+    if (ctx->battleMons[battlerIdTarget].moveEffectFlags & MOVE_EFFECT_MIRACLE_EYE) {
+        if (sTypeEffectiveness[index][1] == TYPE_DARK && sTypeEffectiveness[index][2] == TYPE_MUL_NO_EFFECT) {
+            ret = FALSE;
+        }
+    }
+    
+    return ret;
+}
+
+#ifdef NONMATCHING
+//FIXME: Matches up to regswaps: https://decomp.me/scratch/h06Ji -adrienn
+int ov12_02251D28(BattleSystem *bsys, BATTLECONTEXT *ctx, int moveNo, int moveTypeDefault, int battlerIdAttacker, int battlerIdTarget, int damage, u32 *moveStatusFlags) {
+    int i;
```

Comment:
I personally don't really like the idea of declaring all the vars at the top of a function like this.... looks kinda ugly.... if possible, would you be able to declare them inline with assignments (or as late as possible at least) (provided it still matches, that is)

Reply:
thing is the variable declaration order effects matching a lot of the time due to how the compiler decides to put things in registries, so having them at the top like this is the easiest way to ensure matching